### PR TITLE
Ensure the consistency of targeting evaluation across CPU architectures.

### PR DESCRIPTION
**Pull request by zhiyuanliang-ms**
#403

### CODE REVIEW

**Review by rossgrambo — APPROVED** [review]

File: src/Microsoft.FeatureManagement/Targeting/TargetingEvaluator.cs [hunk 1]
```diff
@@ -348,8 +348,16 @@ private static bool IsTargeted(string contextId, double from, double to)
...
+            // Endianness check ensures the consistency accross different architectures
+            if (!BitConverter.IsLittleEndian)
+            {
+                Array.Reverse(hash, 0, 4);
+            }
+
             //
-            // Use first 4 bytes for percentage calculation
             // Cryptographic hashing algorithms ensure adequate entropy across hash
             uint contextMarker = BitConverter.ToUInt32(hash, 0);
```
Comment (rossgrambo) [inline src/Microsoft.FeatureManagement/Targeting/TargetingEvaluator.cs:362]: I kind of want to explicitly subset the hash to the first 4 bytes just to make things clear. But not needed.

File: src/Microsoft.FeatureManagement/Targeting/TargetingEvaluator.cs [hunk 2]
```diff
@@ -348,8 +348,16 @@ private static bool IsTargeted(string contextId, double from, double to)
                 hash = hashAlgorithm.ComputeHash(Encoding.UTF8.GetBytes(contextId));
             }
 
+
+            //
+            // The first 4 bytes of the hash will be used for percentage calculation
+            // Endianness check ensures the consistency accross different architectures
```
Comment (juniwang) [inline src/Microsoft.FeatureManagement/Targeting/TargetingEvaluator.cs:354]: `accross `. typo?

File: src/Microsoft.FeatureManagement/Targeting/TargetingEvaluator.cs (hunk 1 above)
Comment (zhiyuanliang-ms) [reply src/Microsoft.FeatureManagement/Targeting/TargetingEvaluator.cs:362]: Do you mean:

``` c#
byte[] relevantBytes = new byte[4];

Array.Copy(hash, 0, relevantBytes, 0, 4);


if (!BitConverter.IsLittleEndian)
{
    Array.Reverse(relevantBytes);
}

uint contextMarker = BitConverter.ToUInt32(relevantBytes, 0);
```

I used to implement it like this in [30ec65c43b35ebe46ef7c7327b327b150e7a0f18](https://github.com/microsoft/FeatureManagement-Dotnet/pull/405/commits/30ec65c43b35ebe46ef7c7327b327b150e7a0f18)

Comment (rossgrambo) [reply src/Microsoft.FeatureManagement/Targeting/TargetingEvaluator.cs:362]: Something like that, but I was hoping for something more like:

```csharp
uint contextMarker = BitConverter.ToUInt32(hash.Take(4), 0);
```

Comment (jimmyca15) [reply src/Microsoft.FeatureManagement/Targeting/TargetingEvaluator.cs:362]: That's allocation for no reason.

Comment (rossgrambo) [reply src/Microsoft.FeatureManagement/Targeting/TargetingEvaluator.cs:362]: Reason is for the next developer who's reading it to understand it- the way it is now makes it look like we reverse the first 4 bytes but somehow use all of the bytes to determine a resulting uint.

If we'd rather not add the extra array, maybe we could adjust the comments to make it more clear to me. What if we reordered/adjusted them like this:

```csharp
//
// Cryptographic hashing algorithms ensure adequate entropy across hash
using (HashAlgorithm hashAlgorithm = SHA256.Create())
{
    hash = hashAlgorithm.ComputeHash(Encoding.UTF8.GetBytes(contextId));
}

//
// Endianness check ensures the consistency of targeting evaluation result across different architectures
if (!BitConverter.IsLittleEndian)
{
    Array.Reverse(hash, 0, 4);
}

//
// Only the first 4 bytes of the hash will converted to a uint
uint contextMarker = BitConverter.ToUInt32(hash, 0);
```

Comment (jimmyca15) [reply src/Microsoft.FeatureManagement/Targeting/TargetingEvaluator.cs:362]: Sure comments, but no need to add allocation for it.

**Review by jimmyca15 — APPROVED** [review]

**Pull request merged**